src/mlpack/methods/hmm/hmm_train_main.cpp

```diff
@@ -225,6 +260,29 @@ struct Init
       }
     }
   }
+
+  //! Helper function for Diagonal GMM emission distributions.
+  static void RandomInitialize(vector<DiagonalGMM>& e)
+  {
+    for (size_t i = 0; i < e.size(); ++i)
+    {
+      // Random weights.
+      e[i].Weights().randu();
+      e[i].Weights() /= arma::accu(e[i].Weights());
+
+      // Random means and covariances.
+      for (int g = 0; g < CLI::GetParam<int>("gaussians"); ++g)
+      {
+        const size_t dimensionality = e[i].Component(g).Mean().n_rows;
+        e[i].Component(g).Mean().randu();
+
+        // Generate random diagonal covariance.
+        arma::mat r = arma::diagmat(arma::randu<arma::mat>(dimensionality,
+            dimensionality));
+        e[i].Component(g).Covariance(r * r.t());
```

Comment:
In other models the `r * r.t()` was to guarantee that the covariance generated was positive definite, but if we use a diagonal matrix, we're already guaranteed that---so no need for `r * r.t()` here (just `r` is fine).